### ISSUE TITLE
Typescript 4.5 Beta infinite loop leading to JS heap out of memory

### ISSUE THREAD

**Issue by mmvsk** [colinhacks/zod#689]
An issue appeared with TypeScript 4.5.0-dev.20211001 that made tsc go into an infinite loop leading to an out-of-memory error.

tsconfig.json:

```json
{
  "compilerOptions": {
    "target": "es2020",
    "module": "commonjs",
    "moduleResolution": "node",
    "lib": ["es2020"]
  },
  "exclude": ["node_modules"],
  "include": ["run.ts"]
}
```

package.json:

```json
{
  "type": "commonjs",
  "scripts": {
    "start": "tsc -p ."
  },
  "dependencies": {
    "zod": "^3.9.8"
  },
  "devDependencies": {
    "typescript": "^4.5.0-beta"
  }
```

when using `tsc` directly, the following import cause the issue:

```ts
import { z } from "zod";
```

when using ts-node instead of tsc, this is the minimum necessary code to cause the issue:

```ts
import { z } from "zod";

const schema: z.ZodSchema<string> = z.string();
```

Demo code in: https://github.com/mmvsk/bug-zod-typescript-4.5.0-oct-01

The same code works until TypeScript 4.5.0-dev.20210930, and the issue was probably introduced with the following TypeScript PR: https://github.com/microsoft/TypeScript/pull/41821

**Comment by MikeRippon** [comment]
For reference, there's another issue [here](https://github.com/microsoft/TypeScript/issues/44299), which I think is probably tracking the same issue (?)

**Comment by mmvsk** [comment]
It's not the same, as Zod was working with TS 4.3, 4.4, and 4.5 until 4.5.0-dev.20210930 (included).

There was a pretty large refactoring in TypeScript PR https://github.com/microsoft/TypeScript/pull/41821, and by quickly looking at the diff and the scope of that PR (refactoring recursion-related code) it may very well have introduced this kind of regression.

However, I don't know if Zod is using a faulty behavior of TS that was fixed in 4.5.0-dev.20211001, or if it's a real new TS issue.

**Comment by mmvsk** [comment]
I just tried on a blank setup (Mac M1, without node previously installed), same issue.

Better to figure this out before TS 4.5 is released.
